runtime/io/io.cpp

```diff
@@ -0,0 +1,440 @@
+        case 2:
+          flags = O_WRONLY | O_TRUNC | O_CREAT;
+        case 4:
+          flags = O_WRONLY | O_APPEND | O_CREAT;
+      }
+
+      if (modes & MODE_P) {
+        flags &= (~O_WRONLY) & ~(O_RDONLY);
+        flags |= O_RDWR;
+      }
+
+      if (modes & MODE_E) {
+        flags |= O_CLOEXEC;
+      }
+
+      if (modes & MODE_X) {
+        flags |= O_EXCL;
+      }
+
+      if (modes & MODE_B) {
+      }
+
+      char * f = getTerminatedString(filename);
+      fd = open(f, flags, access);
+    } else {
+      errno = EINVAL;
+      fd = -1;
+    }
+
+    if (-1 == fd) {
+      return getInjErrorBlock();
+    }
+
+    block * retBlock = static_cast<block *>(koreAlloc(sizeof(block) + sizeof(uint64_t)));
+    retBlock->h = header_int();
+    mpz_init_set_si(result, fd);
+    mpz_ptr p = move_int(result);
+    memcpy(retBlock->children, &p, sizeof(mpz_ptr));
+    return retBlock;
+  }
+
+  block * hook_IO_tell(mpz_t i) {
+    if (!mpz_fits_sint_p(i)) {
+      throw std::invalid_argument("Arg too large for int32_t");
+    }
+
+    int fd = mpz_get_si(i);
+    off_t loc = lseek(fd, 0, SEEK_CUR);
+
+    if (-1 == loc) {
+      return getInjErrorBlock();
+    }
+
+    block * retBlock = static_cast<block *>(koreAlloc(sizeof(block) + sizeof(uint64_t)));
+    retBlock->h = header_int();
+    mpz_t result;
+    mpz_init_set_si(result, (long) loc);
+    mpz_ptr p =  move_int(result);
+    memcpy(retBlock->children, &p, sizeof(mpz_ptr));
+    return retBlock;
+  }
+
+  block * hook_IO_getc(mpz_t i) {
+    if (!mpz_fits_sint_p(i)) {
+      throw std::invalid_argument("Arg too large for int32_t");
+    }
+
+    int fd = mpz_get_si(i);
+    char c;
+    ssize_t ret = read(fd, &c, sizeof(char));
+
+
+    if (-1 == ret) {
+      return getInjErrorBlock();
+    }
+
+    block * retBlock = static_cast<block *>(koreAlloc(sizeof(block) + sizeof(uint64_t)));
+    retBlock->h = header_int();
+    mpz_t result;
+    mpz_init_set_si(result, (int) c);
+    mpz_ptr p = move_int(result);
+    memcpy(retBlock->children, &p, sizeof(mpz_ptr));
+    return retBlock;
+  }
+
+  block * hook_IO_read(mpz_t i, mpz_t len) {
+    if (!mpz_fits_sint_p(i) || !mpz_fits_ulong_p(len)) {
+      throw std::invalid_argument("Arg too large");
+    }
+
+    int fd = mpz_get_si(i);
+    size_t length = mpz_get_ui(len);
+
+    auto result = static_cast<string *>(koreAllocToken(sizeof(string) + length));
+    int bytes = read(fd, &(result->data), length);
+
+    if (-1 == bytes) {
+      return getInjErrorBlock();
+    }
+
+    block * retBlock = static_cast<block *>(koreAlloc(sizeof(block) + sizeof(uint64_t)));
+    retBlock->h = header_string();
+    result = static_cast<string *>(koreResizeLastAlloc(result, bytes, length));
+    set_len(result, bytes);
+    memcpy(retBlock->children, &result, sizeof(string *));
+    return retBlock;
+  }
+
+  block * hook_IO_close(mpz_t i) {
+    if (!mpz_fits_sint_p(i)) {
+      throw std::invalid_argument("Arg too large for int32_t");
```

Comment:
as I said, this should be int, not int32_t. Same for the other ones that call sint. ulong should be unsigned long.